### ISSUE TITLE
Application Shells have broken mobile sidebar (double scrollbars)

### ISSUE THREAD

**Issue by vvo** [tailwindlabs/tailwindui-issues#236]
**What component (if applicable)**
- URL for category: https://tailwindui.com/components/application-ui/application-shells/sidebar
- Component name: Light sidebar with light background

**Describe the bug**
Using the default "Light sidebar with light background", testing on iOS Safari, when scrolling the sidebar button stays on top, but if you focus any input in the content then the sidebar button is lost.

This is linked to https://github.com/tailwindui/issues/issues/142

**To Reproduce**
1. Go to https://skhld.csb.app/ with iOS safari
2. Scroll a little bit the page
3. Focus any input
4. The sidebar button disappears/is partially hidden
5. Try to scroll to the top, now the sidebar button is not clickable

**Expected behavior**
Sidebar button should always be at the top, focusing inputs should not change the scroll position.

**Screenshots**
![RPReplay_Final1593593877](https://user-images.githubusercontent.com/123822/86225225-3b1f2800-bb8a-11ea-9d0f-16a3f187c26e.gif)

**Browser/Device (if applicable)**
 - OS: iOS
 - Browser Safari
 - Version 13

**Additional context**
This may be linked to some weird iOS input focus behavior based on the styles of the application shell.

Also, it seems **there are competing scrollbars** on the layouts, some for the whole page, some only for the `<main>` content, see:

![RPReplay_Final1593594006](https://user-images.githubusercontent.com/123822/86225364-699d0300-bb8a-11ea-9f80-523e1e94c4ef.gif)

Which may be the real issue.

cc @jasonlbeggs since you were on the previous issue. This one is maybe more complete, let me know!

**Comment by georgeboot** [comment]
Note that this issue also occurs when you scroll the content section on iOS, because Safari changes the height of the address bar.

When I scroll to the end of the page, I can't see the bottom. Only when I scroll that 2nd bar, I get to see the page bottom. When I scroll up again, I don't see the top: I first have to use the 2nd scroll bar to scroll the last bit.

![image](https://user-images.githubusercontent.com/884482/89211385-07d51c00-d5c2-11ea-8921-36fada22cd1e.gif)

**Comment by jamisuomalainen** [comment]
Can confirm this bug as well.

**Comment by vvo** [comment]
In the end, here's how I fixed this issue: use https://github.com/willmcpo/body-scroll-lock. You have to apply it on the `<main>` and that should do it.

**Comment by georgeboot** [comment]
@vvo sounds like this only blocks scrolling on the `body` element, but it won't decrease the address bars height when scrolling, right?

**Comment by vvo** [comment]
Hey there, any TailwindUI labs person available to help us out? tl;dr; the mobile application layouts are broken, you can't have a single input inside them it will break the navbar (no more available)

**Comment by vvo** [comment]
So I think I found the "right" fix, which is to change the first div of the layouts from:

```html
<div class="h-screen flex overflow-hidden bg-white">
```

to:

```html
<div class="h-full fixed inset-0 flex overflow-hidden bg-white")
```

The tip and explanation comes from this page: https://blog.opendigerati.com/the-eccentric-ways-of-ios-safari-with-the-keyboard-b5aa3f34228d

Along with that, I disable scroll on body at all times (because we have a fixed navbar, and because the "real" scroll is on the `<main>` div).

You can use https://github.com/willmcpo/body-scroll-lock for that

> @vvo sounds like this only blocks scrolling on the body element, but it won't decrease the address bars height when scrolling, right?

Yes, you also need the fix I provide (fixed positioning).

Good luck all!

**Comment by booni3** [comment]
@vvo thanks for this.

It does fix the scrollbar issue... however, it seems to open up other errors within tailwind UI. As an example, the modal now has an off-screen scroll on IOS Safari.

https://tailwindui.com/components/application-ui/overlays/modals

I believe I have fixed this with the below:

```html
<!-- from -->
fixed z-10 inset-0 overflow-y-auto

<!-- to -->
fixed min-h-full inset-0 z-40 overflow-y-auto
```

and...

```html
<!-- from -->
flex items-end justify-center min-h-screen pt-4 px-4 pb-20 text-center sm:block sm:p-0

<!-- to -->
flex items-end justify-center min-h-full pt-4 px-4 pb-20 text-center sm:block sm:p-0
```

Is there any reason this was closed? Feels like having support on iOS would be a basic need for all of the tailwind UI components i.e. added to the core.

**Comment by georgeboot** [comment]
Yeah some of these issues are inherent to using `h-screen`. iOS Safari does that thing where the address bar somewhat shrinks, and that always breaks when using `h-screen`. Basically what happens is that Safari sets the `h-screen` div to be the height of the body with a **shrunken address bar**. This results in some content not being visible, no matter what.

I guess it would be better to only use `h-screen` on `md` and upwards. This however breaks the left navigation bar, however this can be fixed.

**Comment by vvo** [comment]
> Is there any reason this was closed? Feels like having support on iOS would be a basic need for all of the tailwind UI components i.e. added to the core.

I thought this would be an issue that would be seen as important from Tailwind Labs but it seems that they have a lot to do these days! I too think that components should work really well on mobile and for now, as for application shells, this is not the case at all. I guess if more people use Tailwind UI in production on mobile, they will consider this issue with higher importance.

Good luck!

**Comment by adamwathan** [comment]
Wasn’t closed by us, reopening because I agree we should see what we can do to avoid this issue.

**Comment by simonswiss** [comment]
Adding https://github.com/tailwindlabs/tailwindui-issues/issues/313 to the list

**Comment by andresespinosapc** [comment]
@vvo @booni3 can you provide examples of things that don't work in each case? I've made the change in the HTML suggested by @vvo and everything seems to work fine. If i use the `body-scroll-lock` package scroll breaks in modals and dialogs. But even without the change suggested by @booni3 everything seems to work fine, even modals and dialogs.

Any edge case I should be aware of?

**Comment by vvo** [comment]
@reinink Thanks for the very detailed write-up, appreciate it! Next time I'll spin up one of these layouts I know they will work well on mobile.

**Comment by vvo** [comment]
Hey @reinink were all the application shell of Tailwind UI already updated with this new technique? I do not see min-h-full being applied for example. I do see h-full on body and html tags though.

As for my usecase, I had to set h-full on all ancestors of `main` and then set:

```html
<main class="flex-1 overflow-y-auto">...</main>
```

+ also set overflow:hidden on html and body.

This way, the only scroll bar that is ever shown is the one from `<main>`.

Which defeats what you said about body scrolling but somehow I couldn't find a nicer way to achieve what I wanted. Maybe I need to do the opposite: make sure the only scroll is the one from the body. I'll give it another try.

Right now the sidebar layouts of TailwindUI do not adjust to the main content (no scroll is being shown) so I wonder if everything has been updated already.
